### ISSUE TITLE
The driverless PPD generator chooses resolutions for cupsPrintQuality settings badly

### ISSUE THREAD

**Issue by tillkamppeter** [apple/cups#5091]
The PPD generator for driverless printing takes the list of supported resolutions from the IPP attributes, for example from the printer-resolution-supported attribute, and takes the first entry for Draft, the last for High and the one most in the middle for Normal. This works for most printers as the list is sorted, but some have the list not sorted, like the Epson WF 3620 (attributes attached) where Normal gets a ridiculously high resolution.

My suggestion is at least for Normal to use the default resolution, like printer-resolution-default.

In general printer-resolution-* should perhaps be used only in the case of PDF as output format and for any of the raster formats (currently PWG and Apple Raster) the appropriate attributes (pwg-raster-document-resolution-supported, urf-supported with RS...).

Also the actually lowest and highest resolution in any of the lists should be looked for.

Bug report where I have observed this:
https://bugs.launchpad.net/ubuntu/+source/cups/+bug/1712019

Note that it does not matter whether the PPDs are generated by CUPS or cups-filters as most of the PPD generator I have copied over to cups-filters.

IPP attributes of the printer: 
[Epson-WF-3620-ipp-attrs.txt](https://github.com/apple/cups/files/1243347/Epson-WF-3620-ipp-attrs.txt)

**Comment by michaelrsweet** [comment]
Will try to address this in the next 2.2.x update...

**Comment by tillkamppeter** [comment]
Due to your vacation and not knowing when you are back in relation to Ubuntu's Final Freeze for 17.10 and also due to the fact that I have a (slightly improved) copy of the PPD generator in cups-filters, I have worked on this problem and found a nicely working solution (patch attached).
Each raster PDL (PWG Raster and Apple Raster currently in CUPS, PCLm now new in cups-filters) has printer-specific requirements on supported resolutions which are expressed by a PDL-specific IPP attribute supplied by the printer ("pwg-raster-document-resolution-supported", "urf-supported", "pclm-source-resolution-supported"). For a PPD supporting all PDLs (or as many as possible) supported by the printer only resolutions common between the PDLs can be used and PDLs without sufficient resolution info (attribute missing or listed resolutions bogus -> firmware bug) or with inferior maximum resolution compared to the other raster PDLs should be ignored.
Therefore I have opted to investigate the resolutions already before writing the "*cupsFilter2" lines to be able to ignore PDLs with the above-mentioned problems. So already in the beginning I check which PDLs are supported, then go through the PDLs in the order how desirable they are (PDF, PWG Raster, Apple Raster), and on each I check its resolution list and update a list of common resolutions. If their is no problem with the resolution list (attribute exists, at least one correct resolution entry, max resolution at least as high as of the previously checked PDLs) I write the "*cupsFilter2" line for this PDL.
Each resolution list read from as attribute is checked and cleaned: Sorted, duplicates and bogus (at least 1 dimension < 75 dpi) removed (-> firmware bugs). The "600x2dpi" reported by some Brother printers (-> firmware bug) is corrected to "600dpi".
In the end the min, max, and default resolutions are determined and if the attributes do not contain any resolution info (-> firmware bug) we fall back to 300 dpi.
This should fix all the bugs I got reported on driverless printing in Ubuntu or Debian, as they all turned out to be cause by bad resolution attributes:
https://bugs.launchpad.net/ubuntu/+source/cups/+bug/1712019 (PPD generator uses printer-resolution-supported and not urf-supported, printer-resolution-supported not sorted: 600dpi,600x1200dpi,48
00x2400dpi,1200dpi)
https://bugs.debian.org/cgi-bin/bugreport.cgi?bug=868360 (Brother printer reports 600x2dpi as the only resolution for PWG Raster)
Patch with all these changes:
[ppd-generator-resolution-quality.patch.txt](https://github.com/apple/cups/files/1251911/ppd-generator-resolution-quality.patch.txt)

**Comment by tillkamppeter** [comment]
IPP Attributes of Epson WF-3620 (PPD generator uses printer-resolution-supported and not urf-supported, printer-resolution-supported not sorted: 600dpi,600x1200dpi,4800x2400dpi,1200dpi):
[Epson-WF-3620-ipp-attrs.txt](https://github.com/apple/cups/files/1251973/Epson-WF-3620-ipp-attrs.txt)
IPP Attributes of Brother MFC-9340CDW (reports 600x2dpi as the only resolution for PWG Raster):
[Brother-MFC-9340CDW-ipp-attrs.txt](https://github.com/apple/cups/files/1251975/Brother-MFC-9340CDW-ipp-attrs.txt)

**Comment by tillkamppeter** [comment]
This patch solves also Issue #5088 and Issue #5090.
For the latter I have renamed the option cupsPrintQuality to print-quality and its choices to 3, 4, and 5 (only the machine-readable strings in the PPD). This way the print-quality setting in the 5th argument of the call of filters and backends gets set to the user-selected quality and I have tested on the HP DeskJet 2540 that with "High" I get indeed a better print quality (and not because of the resolution, both "Normal" and "High" have 600 dpi).
